src/mono/mono/mini/interp/transform-opt.c

```diff
@@ -481,45 +484,1072 @@ interp_alloc_offsets (TransformData *td)
 	}
 	final_total_locals_size = ALIGN_TO (final_total_locals_size, MINT_STACK_ALIGNMENT);
 
-	// Iterate over all call args locals, update their final offset (aka add td->total_locals_size to them)
-	// then also update td->total_locals_size to account for this space.
-	td->param_area_offset = final_total_locals_size;
+	// Iterate over all call args locals, update their final offset (aka add td->total_locals_size to them)
+	// then also update td->total_locals_size to account for this space.
+	td->param_area_offset = final_total_locals_size;
+	for (unsigned int i = 0; i < td->vars_size; i++) {
+		// These are allocated separately at the end of the stack
+		if (td->vars [i].call_args) {
+			td->vars [i].offset += td->param_area_offset;
+			final_total_locals_size = MAX (td->vars [i].offset + td->vars [i].size, final_total_locals_size);
+		}
+	}
+	td->total_locals_size = ALIGN_TO (final_total_locals_size, MINT_STACK_ALIGNMENT);
+}
+
+/*
+ * DOMINANCE COMPUTATION
+ */
+
+static GString*
+interp_get_bb_links (InterpBasicBlock *bb)
+{
+	GString *str = g_string_new ("");
+
+	if (bb->in_count) {
+		g_string_append_printf (str, "IN (%d", bb->in_bb [0]->index);
+		for (int i = 1; i < bb->in_count; i++)
+			g_string_append_printf (str, " %d", bb->in_bb [i]->index);
+		g_string_append_printf (str, "), ");
+	} else {
+		g_string_append_printf (str, "IN (nil), ");
+	}
+
+	if (bb->out_count) {
+		g_string_append_printf (str, "OUT (%d", bb->out_bb [0]->index);
+		for (int i = 1; i < bb->out_count; i++)
+			g_string_append_printf (str, " %d", bb->out_bb [i]->index);
+		g_string_append_printf (str, ")");
+	} else {
+		g_string_append_printf (str, "OUT (nil)");
+	}
+
+	return str;
+}
+
+static int
+dfs_visit (TransformData *td)
+{
+	int dfs_index = 0;
+	int next_stack_index = 0;
+	td->bblocks = (InterpBasicBlock**)mono_mempool_alloc0 (td->opt_mempool, sizeof (InterpBasicBlock*) * td->bb_count);
+	InterpBasicBlock **stack = (InterpBasicBlock**)g_malloc0 (sizeof (InterpBasicBlock*) * td->bb_count);
+
+	g_assert (!td->entry_bb->in_count);
+	stack [next_stack_index++] = td->entry_bb;
+
+	while (next_stack_index > 0) {
+		// Pop last added element
+		next_stack_index--;
+		InterpBasicBlock *bb = stack [next_stack_index];
+
+		// Process current bblock
+		td->bblocks [dfs_index] = bb;
+		bb->dfs_index = dfs_index++;
+
+		// Push all nodes to process next
+		for (int i = 0; i < bb->out_count; i++) {
+			InterpBasicBlock *out_bb = bb->out_bb [i];
+			if (out_bb->dfs_index == -1) {
+				stack [next_stack_index++] = out_bb;
+				// Mark node as gray so it is not pushed again
+				out_bb->dfs_index = -2;
+			}
+		}
+	}
```

Comment:
There's a comment in the dominator computation that suggests this is meant to compute a reverse post-order, but if I'm reading the code right it is actually computing a pre-order.

Reply:
(I don't think it matters for correctness since the dominator computation iterates to closure anyway, so it probably only impacts the speed of convergence -- just thought I'd point it out.)

Reply:
I'll admit, I didn't put significant thought into these. I just wanted to get dominator computation quickly, so I can focus on the interesting part. I observed in some compiler theory resources, that the normal way is to do a reverse post order, but I think doing this requires reversing a normal traversal, and I wasn't sure why I had to do this in the first place. It is likely that the algorithms for dominator computation and pruned ssa liveness are not converging as fast as they could be in theory. If on CoreCLR side there was more effort put into optimizing these, I would be interested in importing the ideas in the future. It is not clear to me how beneficial it is at this stage.

Reply:
The code for us is here:
https://github.com/dotnet/runtime/blob/9b7aa3d454cce4ff109773512d0b42f4e29ae5d7/src/coreclr/jit/flowgraph.cpp#L3947-L3984
and `fgRunDfs` is here:
https://github.com/dotnet/runtime/blob/9b7aa3d454cce4ff109773512d0b42f4e29ae5d7/src/coreclr/jit/compiler.hpp#L4717-L4796

We represent exceptional flow faithfully in RyuJIT, so the same DFS is used to eliminate dead blocks without needing to special case EH handlers. One of my learnings from working with this stuff is that the faithful handling of EH turns out to be simpler than not representing it and then having to special case it in various places throughout. For SSA, it translates to having to insert some extra phis into handler blocks when we see definitions inside try-regions, and when we enter the try-region.

We don't actually use the reverse post-order for the liveness closure computation today (mainly because we don't have it computed at all points that need to compute liveness today). We do use it for the dominator tree computation, which looks very similar to yours:
https://github.com/dotnet/runtime/blob/9b7aa3d454cce4ff109773512d0b42f4e29ae5d7/src/coreclr/jit/flowgraph.cpp#L5946-L6100
I think it's likely we'll end up switching liveness to use á reverse post-order at some point in the near future. We are gradually moving towards having a reverse post-order of the basic blocks available at all points in the JIT, at which point that change will be a no-brainer.